### PR TITLE
feat: clicks by country and referrer

### PR DESCRIPTION
**Pull request by pilvikala**
Adding two incremental tables to get clicks by country and referrer.

### CODE REVIEW

File: plugin/bitly/client.py [hunk 1]
```diff
@@ -47,3 +49,19 @@ def get_link_clicks(self, link_id: str):
                 f"Failed to get bilink clicks summary for link id '{link_id}': {resp.text}"
             )
         return resp.json()["link_clicks"]
+
+    def get_link_countries_clicks(self, link_id: str, unit: str):
+        resp = self._get(f"bitlinks/{link_id}/countries?unit={unit}&units=1")
+        if resp.status_code != 200:
+            raise Exception(
+                f"Failed to get bilink clicks countries for link id '{link_id}': {resp.text}"
```
Comment (hydratim) [inline plugin/bitly/client.py]: `bilink` or `bitlink`?

File: plugin/bitly/client.py [hunk 2]
```diff
@@ -47,3 +49,19 @@ def get_link_clicks(self, link_id: str):
                 f"Failed to get bilink clicks summary for link id '{link_id}': {resp.text}"
             )
         return resp.json()["link_clicks"]
+
+    def get_link_countries_clicks(self, link_id: str, unit: str):
+        resp = self._get(f"bitlinks/{link_id}/countries?unit={unit}&units=1")
```
Comment (hydratim) [inline plugin/bitly/client.py:62]: Are you not making the units configurable? I suppose that makes the datetime rounding more complicated.

File: plugin/bitly/get_unit_metrics.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+from datetime import datetime, timedelta
+
+
+def get_start_date(unit_reference: str, unit: str) -> str:
+    match unit:
```
Comment (hydratim) [inline plugin/bitly/get_unit_metrics.py:5]: `match` is only available in Python >= 3.10,  yet your `pyproject.toml` specifies `python = "^3.9,<3.13"`.

File: README.md [hunk 4]
```diff
@@ -16,6 +16,9 @@ spec:
     group_id: ${BITLY_GROUP_ID}     # mandatory
     api_token: ${BITLY_API_TOKEN}   # mandatory
     extract_utm: true               # optional. If set, extracts utm_tags from the long_url into separate columns
+    countries_summary_unit: "month" # optional. unit to use to query last 1 {unit} of clicks by a country. Default: month. Values: hour, day, week, month.
+    referrers_summary_unit: "month" # optional. unit to use to query last 1 {unit} of clicks by a referrer. Default: month. Values: hour, day, week, month.
```
Comment (hydratim) [inline README.md]: Comment out optional fields if they are the default state?

File: plugin/client/client.py [hunk 5]
```diff
@@ -15,19 +15,27 @@ class Spec:
...
+        self.validate_summary_unit(self.countries_summary_unit, "countries_summary_unit")
+        self.validate_summary_unit(self.referrers_summary_unit, "referrers_summary_unit")
 
 
 class Client(ClientABC):
     def __init__(self, spec: Spec) -> None:
         self._spec = spec
         self._client = BitlyClient(
-            spec.api_token, spec.group_id, spec.base_url, spec.extract_utm
+            spec.api_token, spec.group_id, spec.base_url, spec.extract_utm, spec.countries_summary_unit, spec.referrers_summary_unit
```
Comment (hydratim) [inline plugin/client/client.py]: Was this formatted by `black`? I would have assumed it'd break the line

File: plugin/plugin.py [hunk 6]
```diff
@@ -39,7 +39,7 @@ def init(self, spec_bytes, no_connection: bool = False):
 
     def get_tables(self, options: plugin.TableOptions) -> List[plugin.Table]:
         all_tables: List[plugin.Table] = [
-            tables.Bitlinks(self._spec.extract_utm),
+            tables.Bitlinks(self._spec.extract_utm, self._spec.countries_summary_unit),
```
Comment (hydratim) [inline plugin/plugin.py:42]: Don't we need to list all of the tables here? or are the others set up as relations only? I've not tested this before.

What happens if you only care about the child table?

File: plugin/bitly/client.py (hunk 2 above)
Comment (pilvikala) [reply plugin/bitly/client.py:62]: Yeah, I thought about it, but then I really did not find any value in this. It does not make sense to run the sync every three days and actually have it working with the incremental tables. To collect historical statistics for every period(unit) It really works only if you run at a higher frequency than your unit and the number of units needs to always be set to `1`. I'd wait until we get feedback or we actually find out we need a bit more flexibility here.

File: plugin/bitly/get_unit_metrics.py (hunk 3 above)
Comment (pilvikala) [reply plugin/bitly/get_unit_metrics.py:5]: good catch!

File: plugin/client/client.py (hunk 5 above)
Comment (pilvikala) [reply plugin/client/client.py]: forgot to run it :)

File: plugin/plugin.py (hunk 6 above)
Comment (pilvikala) [reply plugin/plugin.py:42]: I don't think you can sync a child table only, or can you?

Comment (hydratim) [reply plugin/plugin.py:42]: It's consistent with `Square` so I presume this implementation is correct.

File: plugin/bitly/client.py (hunk 2 above)
Comment (hydratim) [reply plugin/bitly/client.py:62]: Sounds good

**Review by hydratim — APPROVED** [review]

**Pull request merged**